### ISSUE TITLE
mpn/generic/invert.c almost always returns wrong answer; test code doesn't notice

### ISSUE THREAD

**Issue by wbhart** [wbhart/mpir#148]
The code in question is taken from invert.c, for computing a floating point inverse.

http://www.loria.fr/~zimmerma/papers/invert.c

Recall that the function mpn_invert2 is passed a value A = {ap, n} which is normalised (msb set to 1).

It is supposed to return X = B^n + {xp, n} such that AX < B^(2n) <= A(x+1), i.e. a floating point inverse with implicit msb.

Let us ignore the n == 1 and n == 2 blocks and assume they are computed correctly and restrict our attention to the else block (n >= 3).

The first step is to split n into a low and high half:

l = (n - 1)/2
h = n - l

We recursively call mpn_invert2 on the high half of {ap, n} and stick the result into the high half of {xp, n}. For convenience, let us pretend the low half of {xn, n} is set to zero for now.

Note that because of the inequalities the recursive call satisfied, we have that

{ap, n}*(B^n + {xp, n})

may now be greater than or equal to B^2n, since the original inequality only applied to the top half of {ap, n}, not for the whole value.

The next few lines are clearly designed to multiply out this full product and decrement the top half of {xp, n} until {ap, n}*(B^n + {xp, n}) < B^2n.

Most importantly, once this is done, the top half of {xp, n} is now precisely correct. We cannot increase it by 1.

Now looking through the rest of the code, we note that in the generic case, when th != ZERO, we set cy to at least ONE.

The high limbs of {xp, n} are not touched again until the very end, where we have:

mpn_add_1 (xp + l, xp + l, h, cy)

In other words, the high half of {xp, n} is definitely incremented in the generic case.

Thus the final value of X = {xp, n} will not satisfy AX < B^2n as required.

So the question is, why does this pass the test code?

Looking at the function test_invert2, the return value res is conditionally set to 0 (fail) if AX >= B^2n, in the first half of the test.

But in the second half of the test, this value of res is unconditionally overwritten and returned. Thus the first half of the test effectively does nothing.

It looks like we've tried to add code in MPIR to correct another issue with this code, but this certainly doesn't fix the bug mentioned above.

**Comment by wbhart** [comment]
This issue is now resolved in trunk. Paul ZImmermann came up with a nice correction which he describes in his paper, and we have made the necessary adjustments to our code.
